src/com/sun/jna/Memory.java

```diff
@@ -51,9 +48,71 @@
  * @see Pointer
  */
 public class Memory extends Pointer {
-    /** Keep track of all allocated memory so we can dispose of it before unloading. */
-    private static final Map<Memory, Boolean> allocatedMemory =
-            Collections.synchronizedMap(new WeakHashMap<Memory, Boolean>());
+
+    /**
+     * Keep track of all allocated memory so we can dispose of it before
+     * unloading. This is done using a doubly linked list to enable fast
+     * removal of tracked instances.
+     */
+    private static class LinkedReference extends WeakReference<Memory> {
+
+        private LinkedReference next;
+        private LinkedReference prev;
+
+        private LinkedReference(Memory referent) {
+            super(referent);
+        }
+
+        /**
+         * Add the given {@code instance} to the instance tracking.
+         *
+         * @param instance the instance to track
+         */
+        static LinkedReference track(Memory instance) {
+            // keep object allocation outside the syncronized block
+            LinkedReference entry = new LinkedReference(instance);
+
+            synchronized (LinkedReference.class) {
```

Comment:
Idea: Could we not have a `ReentrantLock` guarding `HEAD` here and only lock it for this case? This would permit items later in the list to finalize/dispose themselves without being impacted by adding to the end of the list (which could use that same lock only if we're removing the head item.)

Reply:
I think the synchronized keyword takes care of more things than just preventing n threads entering the same block. Switching to ReentrantLock might lead to unwritten memory effects, see this [link] (https://www.cs.umd.edu/~pugh/java/memoryModel/jsr-133-faq.html#synchronization)

Reply:
I'm not sure that's a problem here, but assuming it is, substitute "synchronize on a different object" in my query above. 

The key point is that all I have to do is make sure `HEAD` doesn't change while adding to the list.  I can do whatever I want with the rest of the list.  Similarly when I'm disposing, I don't care what happens to `HEAD` (except in the rare case that I'm disposing of the head).    

I'm just suggesting having two objects to lock/synchronize on, one for the HEAD and one for everything else.  If you need to dispose the HEAD you can use both objects.

Reply:
Here's my suggestion:
```
    private static LinkedReference HEAD; // the head of the doubly linked list used for instance tracking
    // Use to synchronize operations on the HEAD only
    private static final Object HEAD_LOCK = new Object();

    private static class LinkedReference extends WeakReference<Memory> {

// ...existing code....

        static LinkedReference track(Memory instance) {
            // keep object allocation outside the syncronized block
            LinkedReference entry = new LinkedReference(instance);

            synchronized (HEAD_LOCK) {
                if (HEAD != null) {
                    entry.next = HEAD;
                    HEAD = HEAD.prev = entry;
                } else {
                    HEAD = entry;
                }
            }

            return entry;
        }

        /**
         * Remove the related instance from tracking and update the linked list.
         */
        private void unlink() {
            synchronized (LinkedReference.class) {
                LinkedReference next;

                if (HEAD != this) {
                    // HEAD can't become 'this' so process
                    if (this.prev == null) {
                        // this entry was detached before, e.g. disposeAll was called and finalizers are
                        // running now
                        return;
                    }
                    next = this.prev.next = this.next;
                } else {
                    // HEAD was currently 'this'. Double check lock before changing it
                    synchronized (HEAD_LOCK) {                  
                        if (HEAD == this) {
                            // HEAD is still 'this', update
                            next = HEAD = HEAD.next;
                        } else {
                            // HEAD points to something else now
                            if (this.prev == null) {
                                return;
                            }
                            next = this.prev.next = this.next;
                        }
                    }
                }

// ... existing code ...
```

Reply:
I will update the code (add some comments) and do the benchmarks again. It might take some days until i have time to do this.